File: src/app/handlers/menu.ts
import { Dispatch, MutableRefObject, SetStateAction, useCallback } from "react";
import { STORAGE_KEY } from "../../constants";
import { GameState, createFarmState, createHouseState } from "../../types";
import { ContextMenuState } from "../types";
import { pointWithinPlanet, currentTimeMs } from "../helpers";
import { serializeGameState } from "../../persistence";

type SetContextMenuState = Dispatch<SetStateAction<ContextMenuState | null>>;

interface BuildHouseOptions {
  gameStateRef: MutableRefObject<GameState>;
  aliveCount: number;
  contextMenuState: ContextMenuState | null;
  setContextMenuState: SetContextMenuState;
}

export function useBuildHouseMenuHandler({
  gameStateRef,
  aliveCount,
  contextMenuState,
  setContextMenuState,
}: BuildHouseOptions) {
  return useCallback(
    () => {
      const menuState = contextMenuState;
      if (!menuState) {
        return;
      }

      if (aliveCount < 1) {
        setContextMenuState(null);
        return;
      }

      if (!pointWithinPlanet(menuState.canvasX, menuState.canvasY)) {
        setContextMenuState(null);
        return;
      }

      const state = gameStateRef.current;
      const houseLimit = state.housesBaseCapacity;
      if (houseLimit > 0 && state.houses.length >= houseLimit) {
        setContextMenuState(null);
        return;
      }

      const houseId = state.nextHouseId;
      state.nextHouseId = houseId + 1;
      const builtAt = currentTimeMs();
      state.houses.push(createHouseState(houseId, menuState.canvasX, menuState.canvasY, builtAt));

      try {
        if (typeof localStorage !== "undefined") {
          localStorage.setItem(STORAGE_KEY, serializeGameState(state));
        }
      } catch (error) {
        console.warn("Failed to persist game state after building house", error);
      }

      setContextMenuState(null);
    },
    [aliveCount, contextMenuState, gameStateRef, setContextMenuState],
  );
}

interface BuildFarmOptions {
  gameStateRef: MutableRefObject<GameState>;
  aliveCount: number;
  contextMenuState: ContextMenuState | null;
  setContextMenuState: SetContextMenuState;
}

export function useBuildFarmMenuHandler({
  gameStateRef,
  aliveCount,
  contextMenuState,
  setContextMenuState,
}: BuildFarmOptions) {
  return useCallback(
    () => {
      const menuState = contextMenuState;
      if (!menuState) {
        return;
      }

      if (aliveCount < 10) {
        setContextMenuState(null);
        return;
      }

      if (!pointWithinPlanet(menuState.canvasX, menuState.canvasY)) {
        setContextMenuState(null);
        return;
      }

      const state = gameStateRef.current;
      const farmLimit = state.farmsBaseCapacity;
      if (farmLimit > 0 && state.farms.length >= farmLimit) {
        setContextMenuState(null);
        return;
      }
      const farmId = state.nextFarmId;
      state.nextFarmId = farmId + 1;
      const builtAt = currentTimeMs();
      const lastProducedMs = builtAt - state.farmCropSpawnIntervalMs;
      state.farms.push(
        createFarmState(farmId, menuState.canvasX, menuState.canvasY, builtAt, lastProducedMs),
      );

      const farmBonus = state.farmLifespanBonusPerFarmMs;
      state.settlerMinLifespanMs += farmBonus;
      state.settlerMaxLifespanMs += farmBonus;
      state.settlers.forEach((settler) => {
        if (settler.phase.kind === "Alive") {
          settler.lifespanMs += farmBonus;
        }
      });

      try {
        if (typeof localStorage !== "undefined") {
          localStorage.setItem(STORAGE_KEY, serializeGameState(state));
        }
      } catch (error) {
        console.warn("Failed to persist game state after building farm", error);
      }

      setContextMenuState(null);
    },
    [aliveCount, contextMenuState, gameStateRef, setContextMenuState],
  );
}

File: src/app/handlers/menu.rs
use super::super::ContextMenuState;
use crate::{
    app::point_within_planet,
    constants::STORAGE_KEY,
    types::{GameState, HouseState},
};
use gloo::storage::{LocalStorage, Storage};
use std::{cell::RefCell, rc::Rc};
use yew::{events::MouseEvent, prelude::*, Callback};

pub fn build_house_menu_handler(
    game_state: Rc<RefCell<GameState>>,
    alive_count: UseStateHandle<usize>,
    context_menu_state: UseStateHandle<Option<ContextMenuState>>,
) -> Callback<MouseEvent> {
    Callback::from(move |_| {
        let menu_state = (*context_menu_state).clone();

        if menu_state.is_none() {
            return;
        }

        if *alive_count < 1 {
            context_menu_state.set(None);
            return;
        }

        if let Some(menu) = menu_state {
            if !point_within_planet(menu.canvas_x, menu.canvas_y) {
                context_menu_state.set(None);
                return;
            }

            if let Ok(mut state_ref) = game_state.try_borrow_mut() {
                let house_limit = state_ref.houses_base_capacity as usize;
                if house_limit > 0 && state_ref.houses.len() >= house_limit {
                    context_menu_state.set(None);
                    return;
                }

                let house_id = state_ref.next_house_id;
                state_ref.next_house_id = state_ref.next_house_id.saturating_add(1);
                let built_at = crate::app::current_time_ms();
                state_ref.houses.push(HouseState::new(
                    house_id,
                    menu.canvas_x,
                    menu.canvas_y,
                    built_at,
                ));

                if let Err(error) = LocalStorage::set(STORAGE_KEY, &*state_ref) {
                    log::warn!(
                        "Failed to persist game state after building house: {:?}",
                        error
                    );
                }
            }

            context_menu_state.set(None);
        }
    })
}
